ec/src/hashing/curve_maps/elligator2/mod.rs

```diff
@@ -0,0 +1,268 @@
+use crate::models::twisted_edwards::{MontCurveConfig, TECurveConfig};
+use ark_ff::{Field, One, Zero};
+use ark_std::string::ToString;
+use core::marker::PhantomData;
+
+use crate::{
+    hashing::{map_to_curve_hasher::MapToCurve, HashToCurveError},
+    models::twisted_edwards::{Affine, Projective},
+};
+
+/// Trait defining the necessary parameters for the Elligator2 hash-to-curve method
+/// for twisted edwards curves form of:
+/// `b * y² = x³ + a * x² + x` from [\[WB2019\]]
+/// according to [\[HSSWW22\]]
+///
+/// - [\[BHKL13\]] <http://dx.doi.org/10.1145/2508859.2516734>
+/// - [\[HSSWW22\]] <https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-16>
+
+pub trait Elligator2Config: TECurveConfig + MontCurveConfig {
+    /// An element of the base field that is not a square root see \[BHKL13, Section 5\].
+    /// When `BaseField` is a prime field, [\[HSSWW22\]] mandates that `Z` is the
+    /// non-square with lowest absolute value in the `BaseField` when its elements
+    /// are represented as [-(q-1)/2, (q-1)/2]
+    const Z: Self::BaseField;
+}
+
+/// Represents the Elligator2 hash-to-curve map defined by `P`.
+pub struct Elligator2Map<P: TECurveConfig>(PhantomData<fn() -> P>);
+
+impl<P: Elligator2Config> MapToCurve<Projective<P>> for Elligator2Map<P> {
+    /// Constructs a new map if `P` represents a valid map.
+    fn new() -> Result<Self, HashToCurveError> {
+        // Verifying that U is a non-square
+        if P::Z.legendre().is_qr() {
+            return Err(HashToCurveError::MapToCurveError(
+                "Z should be a quadratic non-residue for the Elligator2 map".to_string(),
+            ));
+        }
+
+        // We assume that the Montgomery curve is correct and  as such we do
+        // not verify the prerequisite for applicability of Elligator2 map
+        Ok(Elligator2Map(PhantomData))
+    }
+
+    /// Map an arbitrary base field element `element` to a curve point.
+    /// Based on
+    /// <https://github.com/zcash/pasta_curves/blob/main/src/hashtocurve.rs>.
+    fn map_to_curve(&self, element: P::BaseField) -> Result<Affine<P>, HashToCurveError> {
```

Comment:
could we be consistent and call it `point` like in the other mappings?

Reply:
Or at least `u` like in RFC 9380

Reply:
I used point in SWU because pasta implementation used point, but it is not a point per se. We do not impose any topology or geometry on the field, so it is more correct to call it an element (which is going to be mapped to a point on the curve which has geometrical properties). So I found it confusing to call the domain element a point when it is not on a curve. As such, I'm happier to change the name of the parameter in SWU and WB to element instead.

Having said that, if you prefer `point` over `element` I'm not going to dwell on it :-) your call.

Reply:
Done in https://github.com/arkworks-rs/algebra/pull/659/commits/354db69df8048c6ec6a0027c11a03b7f9d7bf15d